connector/mqtt/vernemq/k2v-bridge/app/AgentMessenger.js

```diff
@@ -10,59 +10,124 @@ const utils = require('./utils');
 class AgentMessenger {
   /**
    * Creates an Agent Messenger
-   * @access public
-   * @constructor
    *
-   * @param {Object} mqttClient - a mqtt client to connect to VerneMQ broker.
+   * @param {Object} serviceStateManager
+   *
+   * @constructor
    */
-  constructor(mqttClient) {
-    if (!mqttClient) {
-      throw new Error('no MQTT client was passed');
+  constructor(serviceStateManager) {
+    if (!serviceStateManager) {
+      throw new Error('no ServiceStateManager instance was passed');
     }
 
-    this.mqttClient = mqttClient;
-
-    const config = ConfigManager.getConfig('K2V');
-    this.consumeTopicSuffix = config.messenger['consume.topic.suffix'];
+    this.config = ConfigManager.getConfig('K2V');
+    this.consumerTopicSuffix = this.config.messenger['consumer.topic.suffix'];
 
-    this.consumer = new Consumer({
-      ...config.sdk,
-      'kafka.consumer': config.consumer,
-      'kafka.topic': config.topic,
-    });
     this.logger = new Logger('k2v:agent-messenger');
 
+    this.serviceStateManager = serviceStateManager;
+
+    this.consumer = undefined;
     this.wasInitialized = false;
   }
 
   /**
    * Initializes the agent messenger and registers callbacks for incoming messages.
-   * @access public
+   *
+   * @param {Object} mqttClient
+   *
+   * @return {Promise<any>}
+   *
    * @function init
+   * @public
    */
-  init() {
+  async init(mqttClient) {
     if (this.wasInitialized) {
       this.logger.debug('Kafka Consumer already online, skipping its initialization');
       return;
     }
     this.logger.info('Initializing Kafka Consumer...');
-    this.consumer.init().then(() => {
-      const topic = new RegExp(`^.+${this.consumeTopicSuffix.replace(/\./g, '\\.')}`);
 
-      this.consumer.registerCallback(topic, (data) => {
-        this.mqttClient.publishMessage(data);
+    try {
+      this.consumer = new Consumer({
+        ...this.config.sdk,
+        'kafka.consumer': this.config.consumer,
+        'kafka.topic': this.config.topic,
       });
+      await this.consumer.init();
+      const topic = new RegExp(`^.+${this.consumerTopicSuffix.replace(/\./g, '\\.')}`);
 
+      this.consumer.registerCallback(topic, mqttClient.publishMessage.bind(mqttClient));
+
+      this.serviceStateManager.signalReady('kafka');
```

Comment:
Since you pass 'serviceStateManager' as an attribute for the class, why don't you register the shoutdown and heathcheck in the class?

Reply:
or maybe the class receive 2 parameters in the constructor the functions ready and notready , and the rest to be done outside the class?

Reply:
I was doing this, but then @thiagodpf made [this comment](https://github.com/dojot/dojot/pull/1824#discussion_r527821429) that it was not the best thing to do. I don't know which one is the best to be honest.

Reply:
Isn't it weird to pass the functions? The instance already has all of this encapsulated.

Reply:
I also don't know what is best, but passing an instance of the stateManager class to class X and then using a method of class X outside of it passing to stateManager does not seem cool either.
It gets a little messy, something in each place, I don't know ..

Reply:
I did this to avoid the creation of a singleton. I try to avoid this kind of injection we have here (I tried to remove some injections, like the AgentMessenger -> MQTTClient and vice-versa), I don't think it's really nice, but I think singleton is worse than that.

Reply:
ok

Reply:
My comment [here](https://github.com/dojot/dojot/pull/1824#discussion_r527821429) is because it was inside the object's `constructor`...

I will try to exemplify with class AgentMessenger (maybe the example does not apply): 
I imagine that there will only be one object `AgentMessenger`, but let's assume that if more than one object was constructed, the statement `this.serviceStateManager.signalReady('kafka');` would throw an error the second time (If it was inside the `constructor`).

The recommendation is that the class constructor is used only to initialize the attributes of the object to be instantiated, but the above statement was actually doing things unrelated to initializing the `AgentMessenger` _instance_ attributes. 
I was simply "initializing" the `serviceStateManager`, which (in my opinion) should be done in another scope, not within the AgentMessenger's constructor.

But just to exempt myself from guilt, that was another PR :trollface:

Reply:
hahahahaha I think it makes sense, even in this PR too.